File: week3/pascal-triangle.cpp
// Pascal Triangle 
// Given a positive integer N, return the Nth row of pascal's triangle.
// Pascal's triangle is a triangular array of the binomial coefficients formed by summing up the 
// elements of previous row.

#include <bits/stdc++.h>

using namespace std;
#define ll long long

 // } Driver Code Ends
// #define ll long long
class Solution{
public:
    vector<ll> nthRowOfPascalTriangle(int n) {
        // code here
        int i,j;
        ll mod = 1000000007;
        vector<vector<ll>> dp(n,vector<ll>(n,0));
        
        for(i=0;i<n;i++){
        dp[i][0] = 1;
        dp[i][i] =1;
        }
        for(i=2;i<n;i++){
            for(j=1;j<i;j++){
                dp[i][j] = (dp[i-1][j-1]+dp[i-1][j])%mod;
            }
        }
        
        return dp[n-1];
    }
};


// { Driver Code Starts.


void printAns(vector<ll> &ans) {
    for (auto &x : ans) {
        cout << x << " ";
    }
    cout << "\n";
}

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n;
        cin >> n;
        Solution ob;
        auto ans = ob.nthRowOfPascalTriangle(n);
        printAns(ans);
    }
    return 0;
}
  // } Driver Code Ends

File: interviewbit-python/pascal-triangle.py
# Pascal Triangle
# Given numRows, generate the first numRows of Pascal’s triangle.

# Pascal’s triangle : To generate A[C] in row R, sum up A’[C] and A’[C-1] from previous row R - 1.

class Solution:
    # @param A : integer
    # @return a list of list of integers
    def find_bin_coef(self,n):
        res = 1
        v = []
        v.append(res)
        for i in range(1,n+1):
            res = int(res*(n-i+1)/i)
            v.append(res)
        return v

    def solve(self, A):
        res = []
        for i in range(0,A):
            v = self.find_bin_coef(i)
            res.append(v)

            
        return res
